rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java

```diff
@@ -2505,38 +2678,166 @@ private static void processCharSetImpl(REGlobalData gData, RECharSet charSet) {
+        if (!matches) {
+            for (int i = 0; i < charSet.classContents.nonBMPRanges.size(); i += 2) {
+                if (codePoint >= charSet.classContents.nonBMPRanges.get(i)
+                        && codePoint <= charSet.classContents.nonBMPRanges.get(i + 1)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        if (!matches) {
+            for (int encodedProp : charSet.unicodeProps) {
+                if (UnicodeProperties.hasProperty(encodedProp, codePoint)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        if (!matches) {
+            for (int encodedProp : charSet.negUnicodeProps) {
+                if (!UnicodeProperties.hasProperty(encodedProp, codePoint)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        // Apply set operations for 'v' flag
+        if (!charSet.classContents.setOperations.isEmpty()) {
+            for (SetOperation op : charSet.classContents.setOperations) {
+                boolean operandMatches = checkClassContentsMatch(gData, op.operand, codePoint);
+
+                if (op.type == SetOperationType.SUBTRACT) {
+                    // Remove from result if in subtract operand
+                    if (operandMatches) {
+                        matches = false;
+                    }
+                } else if (op.type == SetOperationType.INTERSECT) {
+                    // Keep only if also in intersect operand
+                    matches = matches && operandMatches;
+                }
+            }
+        }
+
+        return matches == charSet.classContents.sense;
+    }
+
+    // Helper method to check if codePoint matches a ClassContents
+    private static boolean checkClassContentsMatch(
+            REGlobalData gData, ClassContents contents, int codePoint) {
+        // Build a temporary RECharSet for the operand
+        RECharSet tempCharSet = new RECharSet(contents, 65536); // Standard BMP length
+
+        // Process and match
+        processCharSet(gData, tempCharSet);
+
+        boolean matches = false;
+
+        if (codePoint <= 0xFFFF) {
+            int byteIndex = codePoint >> 3;
+            if (!(tempCharSet.length == 0
+                    || codePoint >= tempCharSet.length
+                    || (tempCharSet.bits[byteIndex] & (1 << (codePoint & 0x7))) == 0)) {
+                matches = true;
+            }
+        }
+
+        if (!matches && contents.nonBMPCodepoints.contains(codePoint)) {
+            matches = true;
+        }
 
-        for (int i = 0; i < charSet.classContents.nonBMPRanges.size(); i += 2) {
-            if (codePoint >= charSet.classContents.nonBMPRanges.get(i)
-                    && codePoint <= charSet.classContents.nonBMPRanges.get(i + 1)) {
-                return charSet.classContents.sense;
+        if (!matches) {
```

Comment:
There seems to be some overlap between `classMatcher` and `checkClassContentsMatch` - I feel that it is worth trying to keep all the logic inside `classMatcher` without needing a `checkClassContentsMatch` helper. What do you think ?